qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -568,6 +571,74 @@ def __repr__(self) -> str:
         )
 
 
+class ScalableSymbolicPulse(SymbolicPulse):
+    r"""Subclass of :class:`SymbolicPulse` for pulses with scalable envelope.
+
+    Instance of :class:`ScalableSymbolicPulse` behaves the same as an instance of
+    :class:`SymbolicPulse`, but its envelope is assumed to have a scalable form
+    :math:`\text{amp}\times\exp\left(i\times\text{angle}\right)\times\text{F}
+    \left(t,\text{parameters}\right)`,
+    where :math:`\text{F}` is some function describing the rest of the envelope.
+    Parameters `amp` and `angle` are assumed to appear in :attr:`SymbolicPulse.parameters`.
+
+    When two :class:`ScalableSymbolicPulse` objects are equated, instead of comparing
+    `amp` and `angle` individually, only the complex amplitude
+    :math:'\text{amp}\times\exp\left(i\times\text{angle}\right)' is compared.
+    """
+
+    def __init__(
+        self,
+        pulse_type: str,
+        duration: Union[ParameterExpression, int],
+        parameters: Optional[Dict[str, Union[ParameterExpression, complex]]] = None,
+        name: Optional[str] = None,
+        limit_amplitude: Optional[bool] = None,
+        envelope: Optional[sym.Expr] = None,
+        constraints: Optional[sym.Expr] = None,
+        valid_amp_conditions: Optional[sym.Expr] = None,
+    ):
+        if "amp" not in parameters or "angle" not in parameters:
+            raise PulseError("ScalableSymbolicPulse must have parameters amp and angle")
+
+        super().__init__(
+            pulse_type=pulse_type,
+            duration=duration,
+            parameters=parameters,
+            name=name,
+            limit_amplitude=limit_amplitude,
+            envelope=envelope,
+            constraints=constraints,
+            valid_amp_conditions=valid_amp_conditions,
+        )
+
+    def __eq__(self, other: "ScalableSymbolicPulse") -> bool:
+        symbolic_pulse_comparison = super().__eq__(other)
+        if symbolic_pulse_comparison is not True:
+            return symbolic_pulse_comparison
+
+        complex_amp1 = self.amp * np.exp(1j * self.angle)
+        complex_amp2 = other.amp * np.exp(1j * other.angle)
+
+        if isinstance(complex_amp1, ParameterExpression) or isinstance(
+            complex_amp2, ParameterExpression
+        ):
+            if complex_amp1 != complex_amp2:
+                return False
+        else:
+            if not np.isclose(complex_amp1, complex_amp2):
+                return False
+
+        for key in self.parameters:
+            if key not in ["amp", "angle"] and self.parameters[key] != other.parameters[key]:
+                return False
+
+        return True
+
+    # When __eq__ is modified, __hash__ is automatically set to None, and thus needs to be defined
+    # explicitly.
+    __hash__ = SymbolicPulse.__hash__
```

Comment:
Why hash? These pulses are basically mutable and cannot define hash.

Reply:
Well, the `SymbolicPulse` class has a hash function, which is triggered by some tests - which failed (compiler.test_disassembler,compiler.test_assembler). To get those tests to pass, I thought it made sense to have the same hash here.

Reply:
The `__hash__` checks that all the parameters have been assigned first. Since we can't re-bind parameters on a pulse, the pulse is immutable at this point. All the test failures came from one use case -- the `lru_cache` on `_is_amplitude_valid()` which calculates the `Waveform` to check if it exceeds 1. We could consider doing this cached evaluation a different way if we wanted to avoid using `__hash__` in `SymbolicPulse`.

There are a two other cases where it could be nice to hash pulses but it is not currently done. In the qobj assembler, symbolic pulses that are not supported by the backend are converted to arrays of samples. These sample arrays are stored in a dictionary keyed on the hash. The code could instead hash the symbolic pulse and skip the waveform conversion. Since we mainly target pulse gates now, I don't think this helps too much in practice since there is separate deduplication of pulse gate calibrations in the qobj assembler. That pulse gate deduplication is actually the second use case. Currently, it works by converting the schedule into qobj instructions, casting the instructions into strings, and hashing them. Using more tailored hashing methods for different instructions could work better than hashing strings, but it probably woudn't help too much in practice (Python can build and hash strings pretty quickly and users don't tend to use a ton of pulse gates currently).

Reply:
By the way, if we want to support `__hash__` properly, you should override it here. According to the [Python documentation](https://docs.python.org/3/reference/datamodel.html#object.__hash__),

> The only required property is that objects which compare equal have the same hash value

So the hash needs to account for the angle like `__eq__` does  (which might get a bit awkward with floating point error and shifting angle values by 2 pi).

Reply:
Thanks @wshanks for pointing out the missing implementation of the `__hash__`.
I guess the way to do it properly is to use the value of the complex amp with the same level of accuracy for both `__eq__` and `__hash__`.

Reply:
If there's fuzziness in a floating-point comparison somewhere in `__eq__`, you can't implement `__hash__`.  Technically, the only valid return value from `__hash__` is static (i.e. independent of the input[^1]).  Actually coding that is worse than disabling it, because it makes every `dict` operation the worst-case O(n) instead of amortised constant.

Consider trying to implement `__hash__` for a float subclass that has `def __eq__(a, b): return abs(a - b) <= 2**-10`.  Now the numbers `1` and `1 + (2**-10)` are equal, so must have the same hash, which is also true for every number between them too.  But also, `1 + (2**-9)` is equal to `1 + (2**-10)`, so it has to have the same hash as `1` as well.  We can keep going until `2` must have the same hash as `1`, and so on and so on.[^2]

[^1]: except for special values like `inf` and `nan`
[^2]: strictly the absolute comparison I wrote would actually reduce to exact equality at float values of about `2**43`, I just wanted the maths to be simpler than using proper ULP comparisons.

Reply:
Hi @jakelishman ,
Thank you for your comment. I agree that this implementation seems weird. As soon as I pushed this commit I PM’d some people asking them for an input. Obviously, it would have been smart to include you, and better explain myself here.

However, I think under some assumptions it does work (and by all means, correct me if I’m wrong).

First of all, note that I didn’t use a comparison like the one you have in your `__eq__`. I rounded the values before comparing them. For low enough values, the granularity of the round function creates what is essentially a fixed point value. And here comes the first assumption – I assume that the magnitude of the complex amplitude is O(1). By creating a fixed-point-like representation, we have a concrete (numerical uncertainty free) way to both compare and hash.

The second assumption I make, and I really should have posted it here as a question, is that a fixed-point-like representation with high enough resolution is sufficient for all practical purposes – considering the finite precision of the HW. I am guessing `10^-6` is low enough value, but we’d need to figure it if that’s true.
Going back to the example you provided, while `0.4e-7` and `0.3e-7` will be judged equal and hashed the same, for all practical purposes I am guessing they really do behave the same. Note that there is no cascading effect like the one you described, because of the round function.

 I hope this makes my logic clearer. As I said, this does feel weird, but I think it nonetheless works (under the assumptions I provided). I think we should have a discussion about the three options we have – find an agreed upon hashing solution, let go of hashing, or leave in place the non-unique nature of the amp-angle representation.

Reply:
I haven't read the code in detail for what's being doing, but what you're saying sounds like you're effectively binning the data for certain operations, and maybe not for others. You're right that will let you make a valid hash, but it also needs a lot of caution when using the objects, especially if mathematical operations like addition of pulses do not use the same binning behaviour - you can easily end up in the situation where `a1 == a2` and `b1 == b2`, but `a1 + b1 != a2 + b2`, for example.

If the only valid way to use the objects is an the discretised version of the number line, and what you say here implies that you strongly expect _not_ to use the dynamic range of floating-point values, might things be simpler/clearer for developers and users if you use an integer representation of these with a single floating-point value to store the "bin width" scaling? That makes it harder to combine pulses of different precisions, but I think in the hardware model you're trying to describe, that's impossible anyway.

(I have no idea if that is a sensible idea, I just mention it because it seems to make explicit the model the code you describe implies, and doesn't allow people to forget it.)

On the subject of hashing, my view would generally err on the side of caution if there are _any_ questions about whether it makes sense, is efficient, or a user might reasonably mutate parts of the state that contribute to equality in any situation.

Reply:
Yes, you are correct. I effectively bin for some operations and not others. I can't see any immediate and obvious failure point like the examples you described, because there are very few operations you can actually do with the pulses at the moment. However, I agree that this is a risky maneuver. I don't think we can (or want) to bin the data in the pulse itself - firstly because the representation of the pulse is based on the `amp,angle` represenation, while the binning is carried out for the complex amplitude. Furthermore, I suspect actually binning the data would require some more detailed knowledge about the backend, which will limit the scope of Qiskit. In any case, I don't think this is a good way to go just to solve this non-uniqueness issue. If we don't think this is a good enough solution (and I am starting to feel like this is the case), I think we will be forced to ignore this issue, and leave it as a quirk in the code.

Reply:
Eventually, we removed the hashing, and avoided this issue.